docs: add comprehensive generalized wrappers documentation (#595)
cowprotocol/docs: docs/cow-protocol/reference/contracts/periphery/wrapper.mdx

```solidity
function validateWrapperData(
    bytes calldata wrapperData
) external view override;
```

**Requirements:**
- **Deterministic**: Same input must always produce same output. This means you cannot check the timestamp and revert if the order is valid, for example.
- **View function**: Cannot modify state
- **Revert on invalid**: Revert if data is malformed

### Internal Functions (Provided for integrator use)

#### `_next`

Continues the wrapper chain or calls settlement. Handles all parsing and routing automatically.

```solidity
function _next(bytes calldata settleData, bytes calldata remainingWrapperData) internal;
```

**Behavior:**
- Extracts the next target address (first 20 bytes of `remainingWrapperData`)
- Calls the next wrapper via `wrappedSettle()`, or the settlement contract via `settle()` if no wrappers remain

## Calldata Encoding Specification

The `chainedWrapperData`  parameter of `wrappedSettle(settleData, chainedWrapperData)` is specially encoded to minimize bit shuffling overhead:

```
┌──────┬──────────┬──────────┬──────┬──────────┐
│ Len₁ │  Data₁   │  Addr₂   │ Len₂ │  Data₂   │

> │(2 B) │ (wrap1)  │ (20 B)   │(2 B) │ (wrap2)  │

Reviewer: For clarity, since `wrap1` and `wrap2` come out of nowhere.

```suggestion
│ Len₁ │  Data₁   │  Addr₂   │ Len₂ │  Data₂   │
│(2 B) │  (Len₁)  │ (20 B)   │(2 B) │  (Len₂)  │
```

## CowWrapperHelper

A view-only periphery contract for validating and encoding wrapper chains. See the [source code](https://github.com/cowprotocol/euler-integration-contracts/blob/master/src/CowWrapperHelper.sol) for the complete implementation.

### Interface

```solidity
contract CowWrapperHelpers {
    /// @notice A definition for a single call to a wrapper
    struct WrapperCall {
        /// @notice The smart contract that will be receiving the call
        address target;

        /// @notice Any additional data which will be required to execute the wrapper call
        bytes data;
    }

    /// @notice Validates a wrapper chain configuration and builds the properly formatted wrapper data
    /// @param wrapperCalls Array of calls in execution order
    /// @return chainedWrapperData The encoded wrapper data ready to be passed to the first wrapper's wrappedSettle
    function verifyAndBuildWrapperData(
        WrapperCall[] memory wrapperCalls
    ) external view returns (bytes memory chainedWrapperData);
}
```

## Implementation Requirements for Integrators

### Security Requirements

:::warning Critical Requirements

> 

Reviewer: Tag is never closed.

```suggestion
```

### Gas Overhead

Wrapper execution adds gas overhead to settlements.

**Benchmark (EmptyWrapper on Ethereum mainnet):**

| Metric | Value |
|--------|-------|
| With EmptyWrapper | 217,033 gas |
| Without wrapper | 194,761 gas |
| **Overhead** | **22,272 gas (11.4%)** |

**Scaling Factors:**
- Settlement data size (calldata copying)
- Wrapper logic complexity
- Number of nested wrappers

**Methodology:** Single Uniswap V3 WETH→USDC trade. See [services PR #3700](https://github.com/cowprotocol/services/pull/3700).

## Example Implementations

### EmptyWrapper

Minimal wrapper demonstrating the pattern:

```solidity
contract EmptyWrapper is CowWrapper {
    constructor(address settlement) CowWrapper(settlement) {}

    function _wrap(bytes calldata settleData, bytes calldata wrapperData, bytes calldata remainingWrapperData) internal override {
        // No pre-settlement logic
        _next(settleData, remainingWrapperData);  // Pass through
        // No post-settlement logic
    }

    function validateWrapperData(bytes calldata wrapperData)
        external
        view
        override
    {
        // No validation needed
    }
}
```

> 

Reviewer: Similar comment to [this](https://github.com/cowprotocol/docs/pull/595#discussion_r2918177289), we already link to this contract below.

```suggestion
```

### Execution Flow

```mermaid
sequenceDiagram
    participant Solver
    participant Wrapper1
    participant Wrapper2
    participant Settlement

    Solver->>Wrapper1: wrappedSettle(settleData, wrapperData)
    activate Wrapper1
    Note over Wrapper1: Pre-settlement logic
    Wrapper1->>Wrapper2: _next(remainingData)
    activate Wrapper2
    Note over Wrapper2: Pre-settlement logic
    Wrapper2->>Settlement: settle(settleData)
    activate Settlement
    Note over Settlement: Execute trades
    Settlement-->>Wrapper2: Return
    deactivate Settlement
    Note over Wrapper2: Post-settlement logic
    Wrapper2-->>Wrapper1: Return
    deactivate Wrapper2
    Note over Wrapper1: Post-settlement logic
    Wrapper1-->>Solver: Return
    deactivate Wrapper1
```

### Key Design Principles

1. **Abstract**: There are very few limitations on what a wrapper can/can't do around a settlement transaction
2. **Efficient Encoding**: Wrapper-specific data appended to minimize gas overhead
3. **Nested Support**: Multiple wrappers chain by encoding addresses sequentially, allowing CoW orders
4. **Authentication**: Only allowlisted wrappers can call settlement contract

## Implementation

> Developers looking to integrate using wrappers should copy this all-in-one solidity file in your project.

Reviewer 2: in your project --> into their project folders.

Author: aba6279

    Wrapper2->>Settlement: settle(settleData)
    activate Settlement
    Note over Settlement: Execute trades
    Settlement-->>Wrapper2: Return
    deactivate Settlement
    Note over Wrapper2: Post-settlement logic
    Wrapper2-->>Wrapper1: Return
    deactivate Wrapper2
    Note over Wrapper1: Post-settlement logic
    Wrapper1-->>Solver: Return
    deactivate Wrapper1
```

### Key Design Principles

1. **Abstract**: There are very few limitations on what a wrapper can/can't do around a settlement transaction
2. **Efficient Encoding**: Wrapper-specific data appended to minimize gas overhead
3. **Nested Support**: Multiple wrappers chain by encoding addresses sequentially, allowing CoW orders
4. **Authentication**: Only allowlisted wrappers can call settlement contract

## Implementation

Developers looking to integrate using wrappers should copy this all-in-one solidity file into their project as vendored dependency.

It provides a few base contracts which can serve as the foundation for integration:

- **`ICowWrapper`**: Core interface all wrappers must implement
- **`CowWrapper`**: Abstract base contract providing security and utilities that all wrappers should use

> ```solidity

Reviewer: Still suggesting to not copy the code here but just have a link.
- I don't believe we'll keep this up-to-date once a change is made, we're going to forget this dependency exists.
- If I were writing a wrapper, I'd like to have a more authoritative/better tracked source for this critical part of the wrapper and not just "I copied this from the textual docs."

Previous discussion [here](https://github.com/cowprotocol/docs/pull/595#discussion_r2854599329).

Author: alright since no great solution exists right now to have it embedded its just going to be linked as suggested.

Author: [8d1e9a0](https://github.com/cowprotocol/docs/pull/595/commits/8d1e9a017526aa392008d9b4d26ca97da0e59b45)

            // (if the return is less than 32 bytes, the `require` fails and returns revert, verified in test)
            require(
                // forge-lint: disable-next-line(unsafe-typecast)
                returnData.length == 32 && bytes32(returnData) == bytes32(ICowWrapper.wrappedSettle.selector),
                InvalidNextWrapper(nextWrapper)
            );
        }
    }

    /// @notice Helper function to bubble the raw revert result of an address().call invocation.
    /// @param target The contract to call
    /// @param data The data to send
    /// @return returnData The return data if it was successful
    function _callWithBubbleRevert(address target, bytes memory data) internal returns (bytes memory returnData) {
        bool success;
        (success, returnData) = target.call(data);
        if (!success) {
            // Bubble up the revert reason from the next wrapper. Assembly is unfortunately required to do this.
            assembly ("memory-safe") {
                revert(add(returnData, 0x20), mload(returnData))
            }
        }
    }
}
```

### Quick Start Example

Here's a minimal wrapper implementation to use as a starting point:

> ```solidity

Reviewer: I don't like to see large swathes of code in the docs.
I'd just like to make the code in the docs as minimal as possible because the docs are not a dev environment. This is better for maintainability, reproducibility, testing. The problems for a reader:
- Is this code up to date? If not, when was this working? What tooling did it use, like, Solidity version and Foundry version?
- I want to make a change for clarity. Does my code still work after that?
- I'm reviewing this code. Am I the one responsible for making sure the code works?

We can create a new repo, "generalized wrapper examples" that we can properly test and keep updated.

Case in point, it doesn't compile.

```
$ forge build
[⠊] Compiling...
[⠔] Compiling 5 files with Solc 0.8.30
[⠒] Solc 0.8.30 finished in 376.14ms
Error: Compiler run failed:
Error (3656): Contract "MyWrapper" should be marked as abstract.
 --> src/MyWrapper.sol:6:1:
  |
6 | contract MyWrapper is CowWrapper {
  | ^ (Relevant source part starts here and spans across multiple lines).
Note: Missing implementation: 
  --> src/CowWrapper.sol:80:5:
   |
80 |     function name() external view returns (string memory);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Error (4649): Invalid type for argument in modifier invocation. Invalid implicit conversion from address to contract ICowSettlement requested.
 --> src/MyWrapper.sol:7:48:
  |
7 |     constructor(address settlement) CowWrapper(settlement) {}
  |                                                ^^^^^^^^^^
```

Author: This was never intended to be a fully fledged example of a real wrapper, just a segment of code that has the basic pieces to hit the ground running (and apparently missing the `name` function, but resolvable). As an alternative, do you think it would be better to simply document the individual pieces that are required (ex. `_wrap`, `validateWrapperData`, `name`)? I suppose we kind of already go over that in the lower section "virtual functions for integrators"

Author: removed this section because its kind of a duplicate of existing sections which explain the format pretty well.

Author: 42ac22b

---
sidebar_position: 7
---

# Generalized Wrappers

Generalized wrappers are smart contracts that enable custom logic to execute before and after CoW Protocol settlement operations. This reference documents the contract interfaces, implementation patterns, and on-chain behavior of the wrapper system.

## Architecture

### Execution Flow

```mermaid
sequenceDiagram
    participant Solver
    participant Wrapper1
    participant Wrapper2
    participant Settlement

    Solver->>Wrapper1: wrappedSettle(settleData, wrapperData)
    activate Wrapper1
    Note over Wrapper1: Pre-settlement logic

>     Wrapper1->>Wrapper2: _next(remainingData)

Reviewer: Let's show here that `settleData` should be passed along.

Author: 0258e4c

    Wrapper2->>Settlement: settle(settleData)
    activate Settlement
    Note over Settlement: Execute trades
    Settlement-->>Wrapper2: Return
    deactivate Settlement
    Note over Wrapper2: Post-settlement logic
    Wrapper2-->>Wrapper1: Return
    deactivate Wrapper2
    Note over Wrapper1: Post-settlement logic
    Wrapper1-->>Solver: Return
    deactivate Wrapper1
```

### Key Design Principles

1. **Abstract**: There are very few limitations on what a wrapper can/can't do around a settlement transaction
2. **Efficient Encoding**: Wrapper-specific data appended to minimize gas overhead
3. **Nested Support**: Multiple wrappers chain by encoding addresses sequentially, allowing CoW orders
4. **Authentication**: Only allowlisted wrappers can call settlement contract

## Implementation

Developers looking to integrate using wrappers should copy this all-in-one solidity file into their project as vendored dependency.

It provides a few base contracts which can serve as the foundation for integration:

- **`ICowWrapper`**: Core interface all wrappers must implement
- **`CowWrapper`**: Abstract base contract providing security and utilities that all wrappers should use

> ```solidity

Reviewer: I'd remove this and add a link to the Euler project, so we're sure it's kept up-to-date.

Author: I actually wanted to have an effective embed of the code contents here because its useful to see on the page, but unfortunately this doesn't seem to be easily supported by the current docs website markdown setup. I think it may be best to just manually update it. thoughts?

Reviewer: Maybe @anxolin can help here. I think at some point you were able to use code snippets in the docs for a code tutorial, how did you do it?

Reviewer: Moving the discussion [here](https://github.com/cowprotocol/docs/pull/595#discussion_r2918151340) so I don't need to unfurl conversations on GitHub to see open comments.

            // No more wrapper data - we're calling the final settlement contract
            // Verify the settle data has the correct function selector
            require(bytes4(settleData[:4]) == ICowSettlement.settle.selector, InvalidSettleData(settleData));

            // Call the settlement contract directly with the settle data
            (bool success, bytes memory returnData) = address(SETTLEMENT).call(settleData);

            if (!success) {
                // Bubble up the revert reason from the settlement contract
                assembly ("memory-safe") {
                    revert(add(returnData, 0x20), mload(returnData))
                }
            }
        } else {
            // Extract the next wrapper address from the first 20 bytes of wrapperData
            address nextWrapper = address(bytes20(remainingWrapperData[:20]));

            // Skip past the address we just read
            remainingWrapperData = remainingWrapperData[20:];

            // More wrapper data remains - call the next wrapper in the chain
            CowWrapper(nextWrapper).wrappedSettle(settleData, remainingWrapperData);
        }
    }
}
```

### Quick Start Example

> Here's a minimal wrapper implementation to use as a starting point:

Reviewer: Not for this PR, but: it would be extra cool if we were to provide a template repo with just this.

Author: good idea! Lets see how the 2nd usage of wrappers goes and we can see what the patterns are.

        // Revert if data is invalid
    }
}
```

### Wrapper Public Functions

#### `wrappedSettle`

Entry point for wrapper execution. Validates caller authentication and delegates to `_wrap()`--where integrators place their custom logic.

```solidity
function wrappedSettle(
    bytes calldata settleData,
    bytes calldata chainedWrapperData
) external override {
    // Verify caller is an authenticated solver
    require(AUTHENTICATOR.isSolver(msg.sender), NotASolver(msg.sender));

    // Find out how long the next wrapper data is supposed to be
    uint256 nextWrapperDataLen = uint256(uint16(bytes2(chainedWrapperData[0:2])));

    // Delegate to the wrapper's custom logic
    uint256 remainingWrapperDataStart = 2 + nextWrapperDataLen;
    _wrap(
        settleData,
        chainedWrapperData[2:remainingWrapperDataStart],
        chainedWrapperData[remainingWrapperDataStart:]
    );
}
```

**Parameters:**
- `settleData`: Original `GPv2Settlement.settle(...)` calldata
- `chainedWrapperData`: 

#### Chained Wrapper Data Encoding

Each wrapper in the chain consists of:
- 2 bytes: uint16 length of wrapper-specific data
- `length` bytes: wrapper-specific data for this wrapper

> - 20 bytes: address of next wrapper (omitted for the final wrapper)

Reviewer: It's confusing because the actual final wrapper should be included).

```suggestion
- 20 bytes: address of next wrapper (or stop here if no wrappers are left)
```

The final wrapper in the chain omits the next wrapper address and calls SETTLEMENT directly.

> 

Reviewer: It's confusing because the actual final wrapper should be included).

```suggestion
- 20 bytes: address of next wrapper (or stop without an address if no wrappers are left)

When no wrapper is left, the final wrapper in the chain calls SETTLEMENT directly.
```

Example:
```text
[0x0005][0xAABBCCDDEE][0x1234...ABCD][0x0003][0x112233]

> ↑len    ↑data         ↑next wrapper  ↑len   ↑data

Reviewer: For extra clarity, let's spell out the address. Also, spacing typo.
```suggestion
[0x0005][0xAABBCCDDEE][0x1234567890123456789012345678901234567890][0x0003][0x112233]
↑len    ↑data         ↑next wrapper                               ↑len    ↑data
```

    stakeTokens(token, amount);
}
```

### Internal Functions (Provided)

#### `_next`

Continues the wrapper chain or calls settlement. Handles all parsing and routing automatically.

```solidity
function _next(bytes calldata settleData, bytes calldata remainingWrapperData) internal;
```

**Behavior:**
- Reads next wrapper address from `remainingWrapperData`
- Determines if more wrappers exist or if settlement is next
- Makes appropriate call with correct data
- Handles wrapper nesting automatically

**Implementation Details:**
- Extracts next target address (first 20 bytes)
- Separates settle data from remaining wrapper data
- Calls next wrapper via `wrappedSettle()` or settlement via `settle()`

## Calldata Encoding Specification

Wrapper data uses an optimized encoding to minimize gas overhead:

### Structure

```
┌─────────────────────┬──────┬──────────┬──────────┬──────┬──────────┐
│   Settle Calldata   │ Len₁ │  Data₁   │  Addr₂   │ Len₂ │  Data₂   │
│   (to settlement)   │(2 B) │ (wrap1)  │ (20 B)   │(2 B) │ (wrap2)  │
└─────────────────────┴──────┴──────────┴──────────┴──────┴──────────┘

> │<─── settleData ───>│<──────────── wrapperData ─────────────────────>│

Reviewer: Typo in spacing.

```suggestion
└─────────────────────┴──────┴──────────┴──────────┴──────┴──────────┘
│<─── settleData ────>│<─────────── wrapperData ────────────────────>│
```

```

**Components:**
- **Settle Calldata**: `GPv2Settlement.settle(...)` calldata (function selector + args)
- **Len₁**: 2-byte (uint16) length of wrapper 1's data
- **Data₁**: Wrapper 1's custom data
- **Addr₂**: 20-byte address of wrapper 2
- **Len₂**: 2-byte (uint16) length of wrapper 2's data
- **Data₂**: Wrapper 2's custom data

> Note: The settlement contract address is derived from the `SETTLEMENT` immutable and is not included in `wrapperData`.

Reviewer: Wait what are we encoding here? We discussed the wrapper encoding, that's it, `settle` doesn't need to be injected. All of this shouldn't be here right?

Author: Yea I think you are right. I was sort of thinking about it in terms of the call structure as a whole, but as you mention this isnt helpful for understanding the actual encoding pattern, so removing and clarifying what is displayed is for `chainedWrapperData`. 74bbdfa

> ### Data Consumption

Reviewer: Repeating information?

Author: yea I guess its overkill with the existing docs about the behavior of the internal functions

Author: e163784

Each wrapper in the chain:
1. Reads length prefix (2 bytes)
2. Extracts its data (next N bytes)
3. Passes remaining bytes to `_next()`

The `_next()` function:
1. Reads next address (from appropriate offset)
2. Calls next wrapper or settlement
3. Passes remaining data

## CowWrapperHelper

Utility contract for on-chain validation and encoding.

### Interface

```solidity
contract CowWrapperHelper {
    /// @notice Validate and build wrapper data for multiple wrappers
    /// @param wrapperAddresses Array of wrapper contract addresses
    /// @param wrapperDatas Array of wrapper-specific data (parallel to addresses)
    /// @return wrapperData Encoded wrapper data ready for wrappedSettle call
    function verifyAndBuildWrapperData(
        address[] calldata wrapperAddresses,
        bytes[] calldata wrapperDatas
    ) external view returns (bytes memory wrapperData);

> }

Reviewer: Incorrect function signature.

Author: updated call signatures 8726c82

```

### `CowWrapperHelpers`

To aid external integrators in encoding and verifying wrapper data, an additional view-only contract is provided.

> See the [CowWrapperHelper contract source code](https://github.com/cowprotocol/euler-integration-contracts/blob/main/src/CowWrapperHelper.sol) for the complete implementation.

Reviewer: Broken link.

Author: a5fa98c

#### `validateWrapperData`

Validates wrapper-specific data. Must be deterministic and revert on invalid input.

```solidity
function validateWrapperData(
    bytes calldata wrapperData

> ) external view override returns (bytes memory);

Reviewer: Incorrect function signature again. Once is ok but something went wrong with writing this if this is a repeating issue. I'm guessing it's LLM generated, but in case it's still on you to make sure the docs match the actual code. I won't check other function signatures until you confirm you checked them yourself.

```solidity
// ❌ NOT DETERMINISTIC
function validateWrapperData(bytes calldata wrapperData)
    external view override returns (bytes memory)
{
    uint256 deadline = abi.decode(wrapperData, (uint256));
    require(block.timestamp < deadline, "Expired");  // Changes over time!
    return wrapperData;
}

// ✅ DETERMINISTIC
function validateWrapperData(bytes calldata wrapperData)
    external view override returns (bytes memory)
{
    uint256 deadline = abi.decode(wrapperData, (uint256));
    require(deadline > 0, "Invalid deadline");  // Always same result
    return wrapperData;
}
```

In the example above, your `_wrap` code can always reject deadline past expired instead.

#### 5. Defensive Design

Though a solver would be slashed for doing so, there is no hard guarantee wrapper executes even if user specifies it. If wrapper is critical:

**Option A: Make Order Fail Without Wrapper**
```solidity
// Pull required funds in wrapper
function _wrap(bytes calldata wrapperData) internal override {
    vault.withdraw(user, token, amount);  // Funds now in wrapper
    _next(wrapperData);
    // Settlement uses these funds - fails if wrapper didn't execute

> }

Reviewer: This is terrible, the solver can just take the money from the user and not execute the order, leading to loss of funds. This should totally not be here.

Author: sorry, this example should basically be combined with EIP-1271 (which is specified below). Overall I kind of feel like the examples should be removed entirely because its highly case-by-case basis issue, so I went ahead and removed them. 90e0cd6

(btw the cyclical issue really is annoying. I have a couple ways to deal with it, but so far they always involve effectively passing the entire order data to hash into the function. Not efficient, but its doable and we already do it for euler)

```

**Option B: EIP-1271 Authentication**
```solidity
// Order signature only valid if wrapper executed
function isValidSignature(bytes32 hash, bytes memory signature)
    external view returns (bytes4)
{
    require(wrapperExecuted[hash], "Wrapper required");
    return EIP1271_MAGIC;
}

> ```

Reviewer: This is actually extremely annoying and complex to do because of the cyclical dependency on the app data. The wrapper can't generate `hash` without knowing the app data unless doing introspection on the settlement.

Author: agreed, I have a proper solution I think for this, but lets save for another PR. For now removed this example entirely because its highly implementation/need dependent 90e0cd6

Minimal wrapper demonstrating the pattern:

```solidity
contract EmptyWrapper is CowWrapper {
    constructor(address settlement) CowWrapper(settlement) {}

    function _wrap(bytes calldata settleData, bytes calldata wrapperData, bytes calldata remainingWrapperData) internal override {
        // No pre-settlement logic
        _next(settleData, remainingWrapperData);  // Pass through
        // No post-settlement logic
    }

    function validateWrapperData(bytes calldata wrapperData)
        external
        view
        override
    {
        // No validation needed
    }
}
```

### Flash Loan Wrapper

Wrapper coordinating with lending protocol:

```solidity
contract FlashLoanWrapper is CowWrapper {
    ILendingPool public immutable lendingPool;

    constructor(address settlement, address _lendingPool)
        CowWrapper(settlement)
    {
        lendingPool = ILendingPool(_lendingPool);
    }

    function _wrap(bytes calldata settleData, bytes calldata wrapperData, bytes calldata remainingWrapperData) internal override {
        (address asset, uint256 amount) =
            abi.decode(wrapperData, (address, uint256));

        // Store data for use in callback

>         // (In a real implementation, you would store settleData and remainingWrapperData)

Reviewer: Let's drop this section. It feels like this is the most important thing to show and it's omitted. Also, it probably requires some kind of authentication and is complicated to do, and I wouldn't want that we provide a known-broken implementation of a wrapper.

Author: yea looking through all this it seems like it would be better to save these types of examples for later as a deliberate separate PR add

Author: 90e0cd6

        address,
        bytes calldata params
    ) external returns (bool) {
        require(msg.sender == address(lendingPool), "Unauthorized");

        // Execute settlement with borrowed funds
        // (In a real implementation, retrieve stored settleData and remainingWrapperData)
        // _next(settleData, remainingWrapperData);

        // Approve repayment
        IERC20(asset).approve(address(lendingPool), amount + premium);
        return true;
    }

    function validateWrapperData(bytes calldata wrapperData)
        external
        view
        override
    {
        require(wrapperData.length >= 64, "Invalid data length");
        (address asset, uint256 amount) = abi.decode(
            wrapperData,
            (address, uint256)
        );
        require(asset != address(0) && amount > 0, "Invalid parameters");
    }
}
```

## Known Limitations

### Authentication Required

Wrappers need protocol approval, and are not fully permissionless.

### Settlement Upgrades

When settlement contract upgrades, all wrappers must be redeployed. It is recommended to use [Cannon](https://usecannon.com/learn) to make it easy to redeploy your wrappers.

> 

Reviewer: We already discussed limitations somewhere else, no need to do a much shorter versions of it. I also wouldn't mention upgrades like it was something we're happy to do willy-nilly.

```suggestion
```

## Events

No wrapper-specific events are emitted. Wrapper usage can be identified by:

1. Transaction target is wrapper address (not settlement)
2. Function signature is `wrappedSettle(bytes,bytes)` (not `settle(...)`)
3. Analyzing transaction trace for wrapper contract calls

## Resources

### Documentation
- **[Wrapper Concepts](../../../concepts/order-types/wrappers.md)** - High-level overview and use cases
- **[Integration Guide](../../../integrate/wrappers.mdx)** - Implementation guide for order creators and developers
- **[GPv2Settlement](../core/settlement.md)** - Core settlement contract documentation

### Code

> - **[CowWrapper.sol](https://github.com/cowprotocol/euler-integration-contracts/blob/main/src/CowWrapper.sol)** - Abstract contract and interfaces

Reviewer: Broken link.

Author: fixed at same time as other previous link

- **[EmptyWrapper](https://github.com/cowprotocol/euler-integration-contracts)** - Minimal reference implementation

> - **[Euler Integration](https://github.com/cowprotocol/euler-integration-contracts/pull/6)** - Complete production wrapper example

Reviewer: Incorrect links.

Author: fixed 78b02f2
